### PR TITLE
fix: skip clear message when node transition from pending to fail. Fixes #13200

### LINKED ISSUE #13200: Pod doesn't retry when using `"PodInitializing"` in `TRANSIENT_ERROR_PATTERN`
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

When a spot node get evicted, pod has reason/message `"PodInitializing"`
However, specify `"PodInitializing"` in TRANSIENT_ERROR_PATTERN doesn't retry the pod

### Version

latest (https://github.com/argoproj/argo-workflows/commit/669d4a4e986b359dde3cd07e5616be8f713fe23d)

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
Any workflow, but node needs to be evicted during podinitializing
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
timestamp0,                    status: pending, reason: ""
timestamp1,                    status: pending, reason: "PodInitializing"
timestamp2(immediately after), status: fail,    reason: ""
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by tczhao**
Fixes #13200 

### Motivation

Allow retry to use pod message when pod transition from pending to fail.

normally we have
```
timestamp0, status: pending, reason: ""
timestamp1, status: fail,    reason: "e.g. containerd issue"
```

but for podinitializing, the transition are the following
```
timestamp0,                    status: pending, reason: ""
timestamp1,                    status: pending, reason: "PodInitializing"
timestamp2(immediately after), status: fail,    reason: ""
```

### Modifications

This PR fixes the issue, we don't overwrite message with "" when pod transition from pending to fail phase
```
timestamp0,                    status: pending, reason: ""
timestamp1,                    status: pending, reason: "PodInitializing"
timestamp2(immediately after), status: fail,    reason: "PodInitializing"
```

### Verification

Add unit test, test failed
Add changes, test succeed
Also release to our production env for a week and pods able to retry on all `PodInitializing` message when configured in `TRANSIENT_ERROR_PATTERN` without seeing any other issue

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1471,8 +1471,8 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		}
 	}
 
-	// if we are transitioning from Pending to a different state, clear out unchanged message
-	if old.Phase == wfv1.NodePending && new.Phase != wfv1.NodePending && old.Message == new.Message {
+	// if we are transitioning from Pending to a different state (except Fail), clear out unchanged message
+	if old.Phase == wfv1.NodePending && new.Phase != wfv1.NodePending && new.Phase != wfv1.NodeFailed && old.Message == new.Message {
```
Comment (agilgur5) [inline workflow/controller/operator.go:1475]: should this also applied to the `Errored` phase?

**Comment by agilgur5** [comment]
I'm wondering if this might partially help with #11354 and #12572 too 🤔

File: workflow/controller/operator.go (hunk 1 above)
Comment (tczhao) [reply workflow/controller/operator.go:1475]: Make sense

**Comment by tczhao** [comment]
> I'm wondering if this might partially help with #11354 and #12572 too 🤔

I think so, as a workaround solution, use `Message` instead of `ExitCode`

**Review by agilgur5 — APPROVED** [review]
makes sense to me

**Comment by juliev0** [comment]
I see this has been approved. @tczhao do you want to fix the merge conflict, and then it can be merged?

**Pull request merged**